src/rpc/rpc.cc

```diff
@@ -1170,16 +1170,27 @@ future<> server::connection::send_unknown_verb_reply(std::optional<rpc_clock_typ
       });
   }
 
-  future<> server::stop() {
+  future<> server::shutdown() {
+      if (_shutdown) {
+          return make_ready_future<>();
+      }
+
       _ss.abort_accept();
       _resources_available.broken();
       if (_options.streaming_domain) {
           _servers.erase(*_options.streaming_domain);
       }
-      return when_all(_ss_stopped.get_future(),
+      return
```

Comment:
But you still wait for connections to top here. From the commit message I was expecting to see `void shutdown()` that only kicks the process and all the waiting happens in `stop()`.

Reply:
I thought the same, but then changed my mind. The main goal here is to make `messaging_service::shutdown()` stop doing networking communications, just shutting down all sockets (with a syscall) doesn't look sufficient, it should also wait for packets to stop traveling and that's what connection::stop() does. The server::stop() is to destroy the server completely.

I admit that some parts of connection::stop() are better _not_ to be waited in shutdown. If there are some then we'd need `future<> connection::shutdown()` too, but, again, so far I don't see how it will differ from `connection::stop()`.

Does this make sense to you?

Reply:
I need to understand what guaranties you need from the `shutdown()`. If you want to make sure that no rpc handlers issued on this connection are running any longer `connection->stop()` is not enough. What it guaranties is that no new rpcs will be admitted (but they are may be some running already) and no replies will be sent to already admitted once. Is this what you want? What is the use case?

Reply:
> If you want to make sure that no rpc handlers issued on this connection are running any longer connection->stop() is not enough. 

No, that's not the goal here.

> What it guaranties is that no new rpcs will be admitted (but they are may be some running already) and no replies will be sent to already admitted once. Is this what you want?

Yup, that's what I need.

Reply:
> 
> > What it guaranties is that no new rpcs will be admitted (but they are may be some running already) and no replies will be sent to already admitted once. Is this what you want?
> 
> Yup, that's what I need.

Then I can confirm it is what you get with this implementation :) But what is the use case may I ask?

Reply:
> Then I can confirm it is what you get with this implementation :) 

:champagne: 

> But what is the use case may I ask?

Sure! The use-case is described in scylladb/scylla#14031 . The thing is that `storage_service::do_drain()` eventually calls `messaging_service::shutdown()` which is a bit heavyweight -- it, among other things, waits for all RPC handlers to stop running with the help of `_reply_gate` (I added debug for it in #1619 , you're welcome to review and commit it too).  I want to relax `messaging_service::shutdown()` to only isolate the service from the networking -- i.e. call this PR's `rpc::server::shutdown()` -- and do full stopping later, with `rpc::server::stop()`. It's going to look like this -- https://github.com/xemul/scylla/tree/br-messaging-service-split-drain-and-stop (no PR for this branch yet)

Reply:
But should isolation also makes sure that no connection are accepted any longer? I think there is a bug in the current implementation that spilled into this patch as well. Before we stop all connections we need to wait for accept fiber to complete otherwise it may add more connections later. Current code waits for accept in parallel with connection closing and after your patch it does it much later.

Reply:
Hm... The accept looks like

```
    _ss.accept().then([] (auto sk) {
        ... // no co_await-s here
        _conns.emplace(sk);
    });
```
so if `accept()` wasn't aborted and resolved there's no way shutdown could start killing connections before this new one gets emplaced into `_conns`. What am I missing?

Reply:
Ah, it can smp::submit_to connections between shards...